### PR TITLE
RecurrenceRule: Respect leap months

### PR DESCRIPTION
**Pull request by hristost**
This change makes a fix to `Calendar.RecurrenceRule` with regards to leap months and adds a few tests to verify correctness. When constructing the base sequence, we include `isLeapMonth` of the start date in the date components. Before, start dates falling on a leap month would have been treated the same as dates which do fall on the non-leap month before.

### CODE REVIEW

**Comment by hristost** [comment]
@swift-ci please run

File: Tests/FoundationInternationalizationTests/CalendarRecurrenceRuleTests.swift [hunk 1]
```diff
@@ -84,5 +136,5 @@ final class CalendarRecurrenceRuleTests: XCTestCase {
             Date(timeIntervalSince1970: 1730712000.0), // 2024-11-04T01:20:00-0800
         ]
         XCTAssertEqual(results, expectedResults)
-   }
```
Comment (hristost) [inline Tests/FoundationInternationalizationTests/CalendarRecurrenceRuleTests.swift:87]: some unrelated whitespace changes here, these would have happened in one commit or another.

**Comment by hristost** [comment]
@swift-ci please test

**Comment by hristost** [comment]
@swift-ci please test

**Review by itingliu — APPROVED** [review]
